[vscode] support editor/linenumber/context menu

rschnekenbu opened the pull request:
#### What it does

_Adds a missing predefined entry in the menus_
This enables the possibility to contribute menus to editor/linenumber/context entry as defined in vscode 1.78 (see   https://code.visualstudio.com/updates/v1_78#_editorlinenumbercontext-menu)

Fixes #12560

Contributed on behalf of ST Microelectronics

#### How to test 
1. Install following extension:
-  extension: [sample-menu-extension-0.0.3.zip](https://github.com/eclipsesource/theia/files/11755214/sample-menu-extension-0.0.3.zip)
- source: [sample-menu-extension-src.0.0.3.zip](https://github.com/eclipsesource/theia/files/11755220/sample-menu-extension-src.0.0.3.zip)

2. Try to open a menu on any text file, no menu shall open when right-clicking a line number. An error shall also be displayed in the console when the extension is activated. 
3. Update your `theia` repository to use this PR
4. Any text file shall now display at least one menu on line numbers. If line selected is above 11, then a second menu shall also be displayed. This demonstrates the context key editorLineNumber is set correctly. The information displayed when selecting the action shows the current URI for the edited resource, and the line number, as expected by vscode.

5. The build-in github extension shall also start without displaying a comment on a missing menu entry for editor/linenumber/context.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- [x] As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,86 @@
+
+    protected readonly toDisposeOnClosedEditor = new DisposableCollection();
+
+    protected addLineNumberContextMenu(editorWidget: EditorWidget): void {
+        const editor = editorWidget.editor;
+
+        if (editor) {
+            this.toDisposeOnClosedEditor.push(
+                editor.onMouseDown(event => this.handleContextMenu(editor, event))
+            );
+
+            editorWidget.onDispose(() => this.toDisposeOnClosedEditor.dispose());
+        }
+    }
```
nitpick: the extra newlines are unnecessary:
```suggestion
    protected addLineNumberContextMenu(editorWidget: EditorWidget): void {
        const editor = editorWidget.editor;
        if (editor) {
            this.toDisposeOnClosedEditor.push(
                editor.onMouseDown(event => this.handleContextMenu(editor, event))
            );
            editorWidget.onDispose(() => this.toDisposeOnClosedEditor.dispose());
        }
    }
```

rschnekenbu replied:
fixed, thanks for noticing

vince-fugnitto reviewed (commented):
I have some minor comments about the code and formatting.

When testing I noticed some odd behaviours:
- the `lineNumber` is 1 position less than what is selected
- the extension might have a typo in the command registration of `sample-menu-extension.helloWorld` - in vscode this command is never available in the context-menu (as it has no title) while theia displays it

The error in definition is the missing `context` in the `id`:

```json
"contributes": {
  "commands": [
    {
      "command": "sample-menu-extension.context.helloWorld",
      "title": "Sample Extension Editor LineNumber Context"
    },
    {
      "command": "sample-menu-extension.lineGreaterThan10",
      "title": "LineNumber Greater than 10!"
    } 
  ],
  "menus": {
    "editor/lineNumber/context": [
      {
        "command": "sample-menu-extension.helloWorld",
        "icon" : "$(notebook-execute)"
      },
      {
        "command": "sample-menu-extension.lineGreaterThan10",
        "icon" : "$(notebook-execute)",
        "when": "editorLineNumber > 10"
      } 
    ] 
  }
}
```

rschnekenbu commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,86 @@
+import {
+    FrontendApplicationContribution,
+    QuickInputService, ApplicationShell, ContextMenuRenderer
+} from '@theia/core/lib/browser';
```
Thanks for catching this! I fixed this with new commit.

rschnekenbu commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,86 @@
+    @inject(QuickInputService) @optional()
+    protected readonly quickInputService: QuickInputService;
+
+    @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer;
```
I formatted on 2 lines (I find this format more readable) and reordered the injection.

rschnekenbu commented:
Thanks a lot for your review, @vince-fugnitto!

I fixed the formatting issues in a first commit. 
In the second one, I addressed the line number issue, but I am not very pleased on the current fix. The issue comes from the position in MouseTarget, which is a 0-based position. The interesting element is that monaco editor position line number is 1-based, which is kind of confusing here.

I updated the examples, so the basic line number menu is now visible in theia and vscode.
- extension: [sample-menu-extension-0.0.4.zip](https://github.com/eclipse-theia/theia/files/11832331/sample-menu-extension-0.0.4.zip)
- src: [sample-menu-extension-0.0.4-src.zip](https://github.com/eclipse-theia/theia/files/11832334/sample-menu-extension-0.0.4-src.zip)

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,95 @@
+            this.toDisposeOnClosedEditor.push(
+                editor.onMouseDown(event => this.handleContextMenu(editor, event))
+            );
+            editorWidget.onDispose(() => this.toDisposeOnClosedEditor.dispose());
```
Maybe you can dispose this disposable.

rschnekenbu replied:
code has been updated for disposable elements. This fixes also the issue displayed on screencast.

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,95 @@
+export class EditorLineNumberContribution implements FrontendApplicationContribution {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
```
Unused

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,95 @@
+    protected readonly quickInputService: QuickInputService;
+
+    onStart(): void {
+        this.editorManager.onCreated(editor => this.addLineNumberContextMenu(editor));
```
This won't work. If two editors are open, the context menu on the line number works, close one of the editors, and the feature stops working.


https://github.com/eclipse-theia/theia/assets/1405703/adecf173-1831-4971-9f5b-018d7c40610f

rschnekenbu replied:
Each editor contribution is now disposed on its own editor disposal.

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,95 @@
+    protected readonly editorManager: EditorManager;
+
+    @inject(QuickInputService) @optional()
+    protected readonly quickInputService: QuickInputService;
```
Unused

rschnekenbu replied:
removed

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,95 @@
+
+}
+
+function lineNumberFromPosition(position: Position | undefined): number | undefined {
+    // position.line is 0-based line position, where the expected editor line number is 1-based.
+    if (position) {
+        return position.line + 1;
+    }
+    return undefined;
+}
```
That's why Theia has m2p and p2m converters.

rschnekenbu replied:
@kittaakos, thanks a lot for your review! 
As far as I can see, the ProtocolToMonacoConvrter is located on the `monaco` package, where the contribution is currently in the `editor` one. So I cannot rely on it. Is there another converter I am not aware of?

kittaakos replied:
> As far as I can see, the ProtocolToMonacoConvrter is located on the `monaco` package, where the contribution is currently in the `editor` one. So I cannot rely on it.

I see. Thanks for the explanation.

> Is there another converter I am not aware of?

I am unaware of another one, so let's not complicate things before the release. Please go ahead.

(On the other hand, the implementation could be in `monaco`. `event.target.type === MouseTargetType.GUTTER_LINE_NUMBERS` is monaco specific anyway.)

kittaakos replied:
Interesting, I cannot resolve my threads on GitHub anymore. Please mark it as resolved and proceed with the merge. Thank you for the updates, @rschnekenbu

rschnekenbu replied:
Thanks a lot for your review. I rebased my contribution on latest master and solved the conflicts.

kittaakos reviewed (changes requested):
Thank you for the contribution. It does not work for me, please see the screencast. The code can be improved a bit.

rschnekenbu commented:
@kittaakos, thanks for your review! 
I have already updated the contribution and resolved conflicts, I will be just pushing as soon as I have your feedback on m2p converter.

vince-fugnitto commented:
@rschnekenbu please ping me when you address the feedback and conflicts, and I'll happily re-review :+1:

rschnekenbu commented:
@vince-fugnitto, thanks for taking care again of this PR! I rebased and resolved the conflicts, and I addressed the comments from you and Akos. The menu is now disposed by its own editor, so there can be several editors opened and closed, the menu contribution remains for an editor until the editor itself is disposed.

vince-fugnitto reviewed (approved):
The changes look good to me, and work well as well :+1:

kittaakos commented on packages/editor/src/browser/editor-linenumber-contribution.ts:
```diff
@@ -0,0 +1,86 @@
+        if (editor) {
+            const disposables = new DisposableCollection();
+            disposables.push(editor.onMouseDown(event => this.handleContextMenu(editor, event)));
+            editorWidget.disposed.connect(() => disposables.dispose());
```
Hello, thank you for the update. Apologies if I left behind with Theia APIs, but I remember this phosphor way of `connect` requires a `disconnect`. It's missing from this code.

```ts
const callback = () => disposable.dispose();
editorWidget.disposed.connect(callback);
disposables.push(Disposable.create(() => editorWidget.disconnect(callback)));
```

(I wrote the code on GH, it could have errors, but I hope it helps you understand my question.)

rschnekenbu replied:
Thanks @kittaakos, I updated the contribution. That really helps, much appreciated!

kittaakos replied:
The changes look great. I appreciate your time updating the PR multiple times

rschnekenbu commented:
All comments have been addressed, this PR can now be merged. Thanks @vince-fugnitto and @kittaakos for your help :+1:
